src/transaction.rs

```diff
@@ -10,78 +10,129 @@ pub const SIGNED_DATA_OFFSET: usize = size_of::<Signature>();
     /// In the future which key pays the fee and which keys have signatures would be configurable.
-    /// * keys[1] - Typically this is the program context or the recipient of the tokens
-    pub keys: Vec<Pubkey>,
-
-    /// The program code that executes this transaction is identified by the program_id.
-    pub program_id: Pubkey,
+    /// * account_keys[1] - Typically this is the program context or the recipient of the tokens
+    pub account_keys: Vec<Pubkey>,
 
     /// The ID of a recent ledger entry.
     pub last_id: Hash,
 
     /// The number of tokens paid for processing and storage of this transaction.
     pub fee: i64,
 
-    /// Userdata to be stored in the account
-    pub userdata: Vec<u8>,
+    /// Keys indentifying programs in the instructions vector.
+    pub program_keys: Vec<Pubkey>,
+    /// Programs that will be executed in sequence and commited in one atomic transaction if all
+    /// succeed.
+    pub instructions: Vec<Instruction>,
 }
 
 impl Transaction {
-    /// Create a signed transaction from the given `Instruction`.
+    /// Create a signed transaction
     /// * `from_keypair` - The key used to sign the transaction.  This key is stored as keys[0]
-    /// * `transaction_keys` - The keys for the transaction.  These are the program state
+    /// * `account_keys` - The keys for the transaction.  These are the program state
     ///    instances or token recipient keys.
-    /// * `userdata` - The input data that the program will execute with
     /// * `last_id` - The PoH hash.
     /// * `fee` - The transaction fee.
+    /// * `program_keys` - The keys that identify programs used in the `instruction` vector.
+    /// * `instructions` - The programs and their arguments that the transaction will execute atomically
     pub fn new(
         from_keypair: &Keypair,
-        transaction_keys: &[Pubkey],
-        program_id: Pubkey,
-        userdata: Vec<u8>,
+        keys: &[Pubkey],
         last_id: Hash,
         fee: i64,
+        program_keys: Vec<Pubkey>,
+        instructions: Vec<Instruction>,
     ) -> Self {
         let from = from_keypair.pubkey();
-        let mut keys = vec![from];
-        keys.extend_from_slice(transaction_keys);
+        let mut account_keys = vec![from];
+        account_keys.extend_from_slice(keys);
         let mut tx = Transaction {
             signature: Signature::default(),
-            keys,
-            program_id,
+            account_keys,
             last_id,
             fee,
-            userdata,
+            program_keys,
+            instructions,
         };
         tx.sign(from_keypair);
         tx
     }
+    pub fn new_one_program(
+        from_keypair: &Keypair,
+        transaction_keys: &[Pubkey],
+        program_id: Pubkey,
+        userdata: Vec<u8>,
+        last_id: Hash,
+        fee: i64,
+    ) -> Self {
+        let program_keys = vec![program_id];
+        let instructions = vec![Instruction {
+            program_id: 0,
```

Comment:
Need a test to ensure the input `program_id` ends up in both the expected places of the returned tx.